usaspending_api/common/etl/primatives.py

```diff
@@ -0,0 +1,145 @@
+    composed_conditionals = [SQL(template).format(column=Identifier(c), **composed_aliases) for c in columns]
+    return SQL(" or ").join(composed_conditionals)
+
+
+def make_column_list(columns: List[str], alias: Optional[str] = None) -> Composed:
+    """
+    Turn a list of columns into a SQL safe string containing the comma separated list of
+    columns.
+
+    if alias provided
+
+        t.id1, t.id2, t.name
+
+    if no alias provided
+
+        id1, id2, name
+
+    """
+    composed_alias = SQL("") if alias is None else SQL("{}.").format(Identifier(alias))
+    composed_columns = [SQL("{}{}").format(composed_alias, Identifier(c)) for c in columns]
+    return SQL(", ").join(composed_columns)
+
+
+def make_column_setter_list(columns: List[str], alias: str) -> Composed:
+    """
+    Turn a list of columns in a SQL safe string containing a comma separated list of
+    column setters for an update statement.
+
+        name = s.name, description = s.description
+
+    """
+    composed_alias = Identifier(alias)
+    template = "{column} = {alias}.{column}"
+    composed_setters = [SQL(template).format(alias=composed_alias, column=Identifier(c)) for c in columns]
+    return SQL(", ").join(composed_setters)
+
+
+def make_join_conditional(key_columns: List[str], left_alias: str, right_alias: str) -> Composed:
+    """
+    Turn a pair of aliases and a list of key columns into a SQL safe string containing
+    join conditionals ANDed together.
+
+        s.id1 = d.id1 and s.id2 = d.id2
+
+    """
+    composed_aliases = {"left_alias": Identifier(left_alias), "right_alias": Identifier(right_alias)}
+    template = "{left_alias}.{column} = {right_alias}.{column}"
+    composed_conditionals = [SQL(template).format(column=Identifier(c), **composed_aliases) for c in key_columns]
+    return SQL(" and ").join(composed_conditionals)
+
+
+def make_join_excluder_conditional(key_columns: List[str], alias: str) -> Composed:
+    """
+    Turn a list of key columns into a SQL safe string containing join excluder
+    conditionals ANDed together.
+
+        s.id1 is null and s.id2 is null
+
+    """
+    composed_alias = Identifier(alias)
+    return SQL(" and ").join([SQL("{}.{} is null").format(composed_alias, Identifier(c)) for c in key_columns])
+
+
+def make_join_to_table_conditional(key_columns: List[str], alias: str, qualified_table_name: Composed) -> Composed:
+    """
+    Turn an alias, table, and a list of key columns into a SQL safe string containing
+    join conditionals ANDed together.
+
+        d.id1 = public.table1.id1 and d.id2 = public.table1.id2
+
+    """
+    composed_aliases = {"left_alias": Identifier(alias), "right_alias": qualified_table_name}
+    template = "{left_alias}.{column} = {right_alias}.{column}"
+    composed_conditionals = [SQL(template).format(column=Identifier(c), **composed_aliases) for c in key_columns]
+    return SQL(" and ").join(composed_conditionals)
```

Comment:
I would have expected `SQL(" and ".join(composed_conditionals))`

Can you confirm the code is correct performing a .join() on the SQL object instead of creating a SQL object from the string?

Reply:
This is correct.  SQL can do joins much like a string.  Pretty sure I use it this way all over the place.  Now I have to go double check everywhere that I didn't accidentally go the OTHER way anywhere...

Reply:
Ok.  Just confirmed.  Can't do string.join([composables]) or composable.join([strings]).   They throw exceptions.

Reply:
Ah, makes sense. running .join() on a composable using an iterable of composable objects. Got it